beacon-chain/sync/data_columns.go

```diff
@@ -0,0 +1,341 @@
+package sync
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/ethereum/go-ethereum/p2p/enode"
+	"github.com/libp2p/go-libp2p/core"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/v5/beacon-chain/core/peerdas"
+	"github.com/prysmaticlabs/prysm/v5/beacon-chain/db/filesystem"
+	"github.com/prysmaticlabs/prysm/v5/beacon-chain/p2p"
+	"github.com/prysmaticlabs/prysm/v5/beacon-chain/p2p/types"
+	"github.com/prysmaticlabs/prysm/v5/beacon-chain/startup"
+	"github.com/prysmaticlabs/prysm/v5/beacon-chain/sync/verify"
+	"github.com/prysmaticlabs/prysm/v5/beacon-chain/verification"
+	"github.com/prysmaticlabs/prysm/v5/config/params"
+	"github.com/prysmaticlabs/prysm/v5/consensus-types/blocks"
+	"github.com/prysmaticlabs/prysm/v5/consensus-types/interfaces"
+	eth "github.com/prysmaticlabs/prysm/v5/proto/prysm/v1alpha1"
+	"github.com/prysmaticlabs/prysm/v5/runtime/logging"
+	"github.com/prysmaticlabs/prysm/v5/runtime/version"
+	"github.com/sirupsen/logrus"
+)
+
+// RequestDataColumnSidecars sends a data column sidecars by root request to one
+// or more peers that can provide the needed data columns.
+func RequestDataColumnSidecars(
+	ctx context.Context,
+	dataColumns map[uint64]bool,
+	block interfaces.ReadOnlySignedBeaconBlock,
+	blkRoot [32]byte,
+	peers []core.PeerID,
+	clock *startup.Clock,
+	p2p p2p.P2P,
+	ctxMap ContextByteVersions,
+	newColumnsVerifier verification.NewDataColumnsVerifier,
+) ([]blocks.RODataColumn, error) {
+	if len(dataColumns) == 0 {
+		return nil, nil
+	}
+
+	remainingColumns := make(map[uint64]bool, len(dataColumns))
+	for col := range dataColumns {
+		remainingColumns[col] = true
+	}
+
+	sidecars := make([]blocks.RODataColumn, 0, len(dataColumns))
+	maxRetries := 3
```

Comment:
I don't understand the logic of this function.
Why not simply, in a loop, try to retrieve the columns with all peers given by `SelectPeersToFetchDataColumnsFrom(remainingColumns, remainingGoodPeers)` and removing bad peers from `remainingGoodPeers`, and give up when we have no enough good peers remaining?

I don't think we need a `maxRetries`.
With the current implementation, if we reach `maxRetries` but some good peers are still in our list, we return on error and miss the opportunity to use all peers.

Reply:
That makes sense. My initial assumption was that if you're repeatedly not getting back columns from peers that say they should have them, there's probably a bigger problem. But it's cleaner to just retry until the peers run out.